sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeUriBuilder.cs

```diff
@@ -241,6 +242,54 @@ public Uri ToUri()
             return _uri;
         }
 
+        /// <summary>
+        /// Gets the blob Uri.
+        /// </summary>
+        internal Uri ToBlobUri()
+        {
+            // AccountName cannot be null
+            // Host.Length mst be at least 5 character longth than AccountName.Length to account for ".dfs."
+            // Host must begin with "[AccountName]."
+            // The substring of Host immediately following "[AccountName]." must be "dfs."
+            if (AccountName != null
```

Comment:
This seems similar to what we do in UriExtensions.GetAccountName. If it _IpStyleUri is false, then AccountName would only be set if either Dfs/Blob is present in the Uri. Could we use this fact here?

Reply:
Nope, because I can't detect if a IP-style URL is associated with the Blob or DFS endpoint.  I need both a non-null account name, and all the conditions listed below.

Reply:
But doesn't the non-null accountName imply the other conditions are true?
These things are checked in GetAccountNameFromDomain.

Reply:
Ah, I guess it wouldn't help for IpStyle name.

Reply:
Exactly, IP-style URLs don't have "dfs" or "blob" so I can't tell if the user passed me a DFS or Blob URL.  I need to know, because I have to call both the DFS and Blob endpoints with the Data Lake library.

Reply:
So for IP-style URLs, we don't modify the URL, right?
I was suggesting that we could do:

```
if (_!IpStyleUri && !string.IsNullOrEmpty(AccountName)) {  
  AccountName.Replace(Constants.DataLake.DfsUriSuffix, Constants.DataLake.BlobUriSuffix,  
   AccountName.Length+1, Constants.DataLake.DfsUriSuffix.Length);  
}
```
Though this does add a hidden dependency on the fact that our AccountName will only ever be populated for non-ip style Url if we have validated that the dfs/blob subdomain is there.

Reply:
Ya, we could switch to adding the IpStyleUri check, but @tg-msft specifically wanted me to check that the "blob" or "dfs" we are replacing is in the correct spot.  What if the URL myblobaccount.blob.core.windows.net is passed into the ToDfsUri() function?  mydfsaccount.dfs.core.windows.net isn't valid, because we modified the account name.

Reply:
But this check is already done in the GetAccountName method which populates AccountName. If blob/dfs isn't in the correct spot, the AccountName would be empty.

Reply:
"blob" or "dfs" could also be part of the account name.

Reply:
Yes, but we wouldn't replace that because we will start the replace at AccountName.Length+1 as you already have (it looks like the string Replace from the code I pasted doesn't have an overload that takes a start index):
                stringBuilder.Replace(Constants.DataLake.DfsUriSuffix, Constants.DataLake.BlobUriSuffix, AccountName.Length + 1, 3);

Reply:
DataLakeUriBuilder doesn't have a IpStyleUrl property, and when I call IsHostIPEndPointStyle(_url), I'm getting multiple unit tests failing with null points.

I think we should leave this as is, it is pretty robust.

Reply:
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeUriBuilder.cs#L31

Reply:
I think calling `GetAccountNameFromDomain` might be the easiest approach.  See my comment below on the `ToDfsUri` method.

Reply:
_uri.GetAccountNameFromDomain() results in a null pointer, because _uri is null until you call ToUri()